sm: launcher: implement rootfs runtime

mlohvynenko opened the pull request.

mykola-kobets-epam reviewed (approved):
`Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>`

MykolaSuperman reviewed (approved):
`Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>`

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -0,0 +1,401 @@
+    auto status = FillInstanceStatus();
+
+    switch (const auto action = ReadAction(); action.GetValue()) {
+    case ActionTypeEnum::eDoUpdate: {
```
empty lines between cases

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -0,0 +1,401 @@
+
+    LOG_DBG() << "Getting rootfs runtime info";
+
+    runtimeInfo = static_cast<const RuntimeInfo&>(mConfig);
```
mRuntimeID should consist of runtimeType + nodeID

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.hpp:
```diff
@@ -0,0 +1,174 @@
+#include "sm/launcher/runtimes/config.hpp"
+#include "sm/launcher/runtimes/imagemanager.hpp"
+
+namespace aos::sm::launcher::rootfs {
```
should be in `aos::sm::launcher` namespace

mlohvynenko committed:
sm: launcher: utils: handle empty units list in SystemdUpdateChecker

Do not treat empty units list as failure or inactive state.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

al1img commented on src/sm/config/config.cpp:
```diff
@@ -150,4 +153,20 @@ Error ParseConfig(const std::string& filename, Config& config)
     return ErrorEnum::eNone;
 }
 
+Error ParseConfig(const Poco::JSON::Object::Ptr& jsonObj, launcher::RootfsConfig& config)
```
Rootfs config should be parsed in rootfs runtime

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -16,41 +31,82 @@ namespace aos::sm::launcher {
-    LOG_DBG() << "Start runtime";
+    std::lock_guard lock {mMutex};
+
+    LOG_DBG() << "Starting rootfs runtime" << Log::Field("version", mCurrentVersion);
```
"Start runtime" it will be in rootfs log module

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -59,18 +115,63 @@ Error RootfsRuntime::GetRuntimeInfo(RuntimeInfo& runtimeInfo) const
-    (void)status;
+    std::lock_guard lock {mMutex};
+
+    LOG_DBG() << "Starting instance" << Log::Field("ident", static_cast<const InstanceIdent&>(instance))
```
Start instance

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -59,18 +115,63 @@ Error RootfsRuntime::GetRuntimeInfo(RuntimeInfo& runtimeInfo) const
 Error RootfsRuntime::StopInstance(const InstanceIdent& instance, InstanceStatus& status)
 {
-    (void)status;
+    LOG_DBG() << "Stopping instance" << Log::Field("ident", instance);
```
Stop instance

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -117,4 +270,341 @@ Error RootfsRuntime::CreateRuntimeInfo(const std::string& runtimeType, const Nod
+{
+    const auto action = ReadAction();
+
+    LOG_DBG() << "Processing rootfs update action" << Log::Field("action", action.ToString().CStr());
```
"Process

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -117,4 +270,341 @@ Error RootfsRuntime::CreateRuntimeInfo(const std::string& runtimeType, const Nod
+        return AOS_ERROR_WRAP(err);
+    }
+
+    Poco::Process::Args args;
```
There is common::utils::ExecCommand or check common::utils::UnpackTarImage if it suits for you

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -16,41 +31,82 @@ namespace aos::sm::launcher {
-    if (auto err = CreateRuntimeInfo(config.mType, *nodeInfo); !err.IsNone()) {
-        return err;
+    if (auto er = fs::MakeDirAll(mRootfsConfig.mWorkingDir.c_str()); !er.IsNone()) {
+        LOG_DBG() << "Failed to create working dir" << Log::Field("dir", mRootfsConfig.mWorkingDir.c_str());
```
Remove this log

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -16,41 +31,82 @@ namespace aos::sm::launcher {
-    LOG_DBG() << "Get runtime info";
+    std::lock_guard lock {mMutex};
+
+    LOG_DBG() << "Getting rootfs runtime info";
```
`"Get runtime info"`

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -96,15 +197,67 @@ Error RootfsRuntime::GetInstanceMonitoringData(
+
+    std::ifstream versionFile(mRootfsConfig.mVersionFilePath);
+    if (!versionFile.is_open()) {
+        return Error(ErrorEnum::eNotFound, "Version file not found");
```
error message from small letter

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -96,15 +197,67 @@ Error RootfsRuntime::GetInstanceMonitoringData(
+    Poco::StringTokenizer tokenizer(versionFileContent, "=", Poco::StringTokenizer::TOK_TRIM);
+
+    if (tokenizer.count() != 2 || tokenizer[0] != "VERSION") {
+        return Error(ErrorEnum::eInvalidArgument, "Invalid version file format");
```
ditto

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -117,4 +270,341 @@ Error RootfsRuntime::CreateRuntimeInfo(const std::string& runtimeType, const Nod
+        std::filesystem::remove(entry.path(), ec);
+        if (ec) {
+            LOG_ERR() << "Failed to remove update artifact" << Log::Field("path", entry.path().c_str())
+                      << Log::Field("error", ec.message().c_str());
```
"err"

al1img commented on src/sm/launcher/runtimes/rootfs/rootfs.cpp:
```diff
@@ -117,4 +270,341 @@ Error RootfsRuntime::CreateRuntimeInfo(const std::string& runtimeType, const Nod
+        std::filesystem::remove(path, ec);
+        if (ec) {
+            LOG_ERR() << "Failed to remove action file" << Log::Field("path", path.c_str())
+                      << Log::Field("error", ec.message().c_str());
```
ditto

consider to creatre local function that converts ec to Aos errror

al1img commented on src/sm/launcher/runtimes/rootfs/tests/rootfs.cpp:
```diff
@@ -21,33 +39,448 @@ namespace {
-    static void SetUpTestSuite() { tests::utils::InitLog(); }
+    void SetUp() override
+    {
+        tests::utils::InitLog();
```
Setup log in: `static void SetUpTestSuite() { tests::utils::InitLog(); }`

al1img commented on src/sm/config/config.cpp:
```diff
@@ -150,4 +153,20 @@ Error ParseConfig(const std::string& filename, Config& config)
+    try {
+        common::utils::CaseInsensitiveObjectWrapper object(jsonObj);
+
+        config.mWorkingDir      = object.GetValue<std::string>("workingDir", cDefaultRootfsRuntimeDir);
```
We have root "workingDir", probably we need to rename this field to something else.

mlohvynenko replied:
agreed to keep as it is

al1img commented on src/sm/config/config.cpp:
```diff
@@ -150,4 +153,20 @@ Error ParseConfig(const std::string& filename, Config& config)
+
+        config.mWorkingDir      = object.GetValue<std::string>("workingDir", cDefaultRootfsRuntimeDir);
+        config.mVersionFilePath = object.GetValue<std::string>("versionFilePath", cDefaultRootfsVersionFile);
+        config.mServicesToCheck = common::utils::GetArrayValue<std::string>(
```
healthCheckServices?

mlohvynenko replied:
renamed

mlohvynenko committed:
sm: config: set empty runtime config objects

Currently, if config tag is missing in runtime config,
json object pointer remains null. This patch simplifies
futher usage by setting empty json object in such cases.

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
sm: launcher: implement rootfs runtime

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
launcher: runtimes: utils: log received systemd unit status

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

mlohvynenko committed:
sm: launcher: config: pass workingDir to runtime configs

Signed-off-by: Mykhailo Lohvynenko <mykhailo_lohvynenko@epam.com>
Reviewed-by: Mykola Kobets <mykola_kobets@epam.com>
Reviewed-by: Mykola Solianko <mykola_solianko@epam.com>
Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>

al1img reviewed (approved):
`Reviewed-by: Oleksandr Grytsov <oleksandr_grytsov@epam.com>`
